### PR TITLE
Add restart module

### PR DESCRIPTION
**Pull request by inducer**
This is a version of #140 that uses https://github.com/inducer/meshmode/pull/99 to remove  a little bit of array context gymnastics from restart.

The diff vs that branch is here: https://github.com/illinois-ceesd/mirgecom/compare/restart-prototype...restart-prototype-v2

cc @MTCam @anderson2981

### CODE REVIEW

**Comment by MTCam** [comment]
We need to port this to a more relevant example in *MIRGE-Com*, and move it away from the old crusty EagerDiscretization. I think we can just roll in an example that looks like @anderson2981's nozzle or so and kill multiple birds with a single projectile.

**Comment by MTCam** [comment]
This is ready for merge as soon as you're a comfy @inducer.

**Comment by inducer** [comment]
LGTM. Can't approve, since it's (nominally) my PR, apparently.

**Comment by MTCam** [comment]
Sorry, just wanted to update the wave-eager-mpi restart example since it demonstrates `restart_order != saved_order`.  I'll make sure it works and submit a review on it.

**Review by MTCam — APPROVED** [review]
Yah, itworks.

**Comment by MTCam** [comment]
Okie for merge @inducer?

File: mirgecom/restart.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    """Pickle the simulation data into a file for use in restarting."""
+    rank = 0
+    if comm:
+        rank = comm.Get_rank()
+    if rank == 0:
+        import os
+        rst_dir = os.path.dirname(filename)
+        if rst_dir:
+            if not os.path.exists(rst_dir):
+                os.makedirs(rst_dir)
```
Comment (inducer) [inline mirgecom/restart.py:54]:
```suggestion
            os.makedirs(rst_dir)
```
(The `if` is redundant.)

**Review by inducer — COMMENTED** [review]
LGTM. One minor suggestion.

File: mirgecom/restart.py (hunk 1 above)
Comment (matthiasdiener) [reply mirgecom/restart.py:54]: Should it be `os.makedirs(rst_dir, exist_ok=True)` ?

**Pull request merged**